src/ILogger/ApplicationInsightsLoggingBuilderExtensions.cs

```diff
@@ -0,0 +1,54 @@
+﻿// -----------------------------------------------------------------------
+// <copyright file="ApplicationInsightsLoggingBuilderExtensions.cs" company="Microsoft">
+// Copyright (c) Microsoft Corporation. 
+// All rights reserved.  2013
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Microsoft.Extensions.Logging
+{
+    using System;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.Extensions.DependencyInjection;
+    using Microsoft.Extensions.DependencyInjection.Extensions;
+    using Microsoft.Extensions.Logging.ApplicationInsights;
+
+    /// <summary>
+    /// Extensions methods to add and configure application insights logger.
+    /// </summary>
+    public static class ApplicationInsightsLoggingBuilderExtensions
+    {
+        /// <summary>
+        /// Adds an ApplicationInsights logger named 'ApplicationInsights' to the factory.
+        /// </summary>
+        /// <param name="builder">The <see cref="ILoggingBuilder"/> to use.</param>
+        public static ILoggingBuilder AddApplicationInsights(this ILoggingBuilder builder)
+        {
+            return builder.AddApplicationInsights((applicationInsightsOptions) => { });
+        }
+
+        /// <summary>
+        /// Adds an ApplicationInsights logger named 'ApplicationInsights' to the factory.
+        /// </summary>
+        /// <param name="builder">The <see cref="ILoggingBuilder"/> to use.</param>
+        /// <param name="configureApplicationInsightsOptions">Action to configure ApplicationInsights logger.</param>
+        public static ILoggingBuilder AddApplicationInsights(
+            this ILoggingBuilder builder,
+            Action<ApplicationInsightsLoggerOptions> configureApplicationInsightsOptions)
+        {
+            if (configureApplicationInsightsOptions == null)
+            {
+                throw new ArgumentNullException(nameof(configureApplicationInsightsOptions));
+            }
+
+            // This ensures that we don't override the TelemetryClient if one was added using .AddApplicationInsightsTelemetry()
```

Comment:
Who’s adding the telemetry client?

Reply:
No one is adding TelemetryClient per say. TelemetryConfiguration is what is added using the existing ASPNetCore extension methods (in ASPNetCore package). However all of this is true provided the user sets up everything correctly. There are chances AI pipeline might not be set correctly (developers might choose to not have it running on your local machine or something similar). For such cases, I am adding a dummy Configuration which is used to create a TelemtryClient which runs the full AI pipeline, but does not send the telemetry outside of the machine.

LoggerProvider takes TelemetryConfiguration and creates TelemetryClient from it and then passes this to every Logger instance.

There is a rather interesting side case as well. ILogger instance can be asked for during the Startup.cs call. However at this point if during Webhost setup, developer did not call .UseApplicationInsights method, all logs emiited during Startup.cs are lost since AI key was not yet setup. Additionally since .UseApplicationInsights method does not allow for adding additional sinks, telemetryInitializers or processors telemetry might be a lot lacking. However once the ConfigureServices and Configure call is made the full pipeline will be created and the second instance of ApplicationInsightsLoggerProvider will be shipping logs to AI and running full pipelines.

But all of this is existing and this code and not change any of that. The console sample at the very top gives a brief idea of how the pipeline comes into effect.